File: web/src/domain/__tests__/ufReajuste.test.ts
import { describe, expect, it } from "vitest";
import { money } from "../money.ts";
import { of } from "../dates.ts";
import { convert } from "../ufReajuste.ts";
import { expectWithin, num } from "../testing.ts";

const uf = (date: string, value: string) => ({ date, value: money(value) });

describe("uf reajuste", () => {
  it("restates by the ratio of the two UF values", () => {
    const r = convert(money("1000"), uf(of(2020, 1, 1), "28000.00"), uf(of(2021, 1, 1), "29000.00"));

    expectWithin(num(r.factor), 1e-6, 29000 / 28000);
    expectWithin(num(r.adjustedAmount), 0.5, (1000 * 29000) / 28000);
    expectWithin(num(r.ufUnits), 0.0001, 1000 / 28000);
  });

  it("four thousand pesos of january 1990 restate to about thirty thousand", () => {
    const r = convert(money("4000"), uf(of(1990, 1, 1), "5435.28"), uf(of(2026, 9, 5), "40880.36"));

    expectWithin(num(r.adjustedAmount), 1, 30085);
    expectWithin(num(r.factor), 0.0001, 7.5213);
    expect(r.days).toBe(13396);
    expectWithin(num(r.annualisedPct), 0.05, 5.66);
  });

  it("the same day is a no-op", () => {
    const day = uf(of(2020, 5, 5), "28000.00");
    const r = convert(money("1234"), day, day);

    expectWithin(num(r.adjustedAmount), 0.5, 1234);
    expectWithin(num(r.factor), 1e-9, 1);
    expectWithin(num(r.variationPct), 1e-9, 0);
    expectWithin(num(r.annualisedPct), 1e-9, 0);
    expect(r.days).toBe(0);
  });

  it("converting forward and back returns the original amount", () => {
    const a = uf(of(1995, 6, 15), "12000.00");
    const b = uf(of(2020, 6, 15), "28000.00");

    const forward = convert(money("100000"), a, b);
    const back = convert(forward.adjustedAmount, b, a);

    expectWithin(num(back.adjustedAmount), 1, 100000);
  });

  it("going backwards in time shrinks the amount", () => {
    const r = convert(money("30000"), uf(of(2026, 9, 5), "40880.36"), uf(of(1990, 1, 1), "5435.28"));

    expect(num(r.adjustedAmount)).toBeLessThan(30000);
    expect(r.variationPct.cmp(0)).toBeLessThan(0);
    expect(r.days).toBe(-13396);
  });

  it("the annual rate compounds back to the factor", () => {
    const r = convert(money("1000"), uf(of(2000, 1, 1), "15000.00"), uf(of(2020, 1, 1), "28000.00"));
    const rebuilt = Math.pow(1 + num(r.annualisedPct) / 100, r.days / 365.25);

    expectWithin(rebuilt, 0.01, num(r.factor));
  });

  it("two dates one day apart give different results", () => {
    const base = uf(of(2026, 9, 1), "40875.09");
    const a = convert(money("1000000"), base, uf(of(2026, 9, 4), "40879.04"));
    const b = convert(money("1000000"), base, uf(of(2026, 9, 5), "40880.36"));

    expect(b.adjustedAmount.cmp(a.adjustedAmount)).toBeGreaterThan(0);
  });

  it("large amounts do not drift from float rounding", () => {
    const r = convert(money("1000000000"), uf(of(2020, 1, 1), "28000.00"), uf(of(2020, 1, 2), "28001.00"));

    expectWithin(num(r.adjustedAmount), 1, (1_000_000_000 * 28001) / 28000);
  });

  it("a non-positive UF value is refused rather than dividing by zero", () => {
    expect(() => convert(money("1000"), uf(of(2020, 1, 1), "0"), uf(of(2021, 1, 1), "1"))).toThrow();
    expect(() => convert(money("1000"), uf(of(2020, 1, 1), "1"), uf(of(2021, 1, 1), "-1"))).toThrow();
  });
});

File: app/src/test/java/cl/tickers/app/domain/UfReajusteTest.kt
package cl.tickers.app.domain

import cl.tickers.app.domain.engine.UfReajuste
import cl.tickers.app.domain.model.DatedValue
import com.google.common.truth.Truth.assertThat
import org.junit.Assert.assertThrows
import org.junit.Test
import java.math.BigDecimal
import java.time.LocalDate

class UfReajusteTest {

    private fun uf(date: String, value: String) = DatedValue(LocalDate.parse(date), BigDecimal(value))

    @Test
    fun `an amount is restated by the ratio of the two UF values`() {
        val result = UfReajuste.convert(
            amount = BigDecimal("1000"),
            from = uf("2020-01-01", "28000.00"),
            to = uf("2021-01-01", "29000.00"),
        )

        assertThat(result.factor.toDouble()).isWithin(1e-6).of(29000.0 / 28000.0)
        assertThat(result.adjustedAmount.toDouble()).isWithin(0.5).of(1000.0 * 29000 / 28000)
        assertThat(result.ufUnits.toDouble()).isWithin(0.0001).of(1000.0 / 28000.0)
    }

    /** The case the screen opens on. */
    @Test
    fun `four thousand pesos of january 1990 restate to about thirty thousand`() {
        val result = UfReajuste.convert(
            amount = BigDecimal("4000"),
            from = uf("1990-01-01", "5435.28"),
            to = uf("2026-09-05", "40880.36"),
        )

        assertThat(result.adjustedAmount.toDouble()).isWithin(1.0).of(30085.0)
        assertThat(result.factor.toDouble()).isWithin(0.0001).of(7.5213)
        assertThat(result.days).isEqualTo(13396L)
        assertThat(result.annualisedPct.toDouble()).isWithin(0.05).of(5.66)
    }

    @Test
    fun `the same day is a no-op`() {
        val day = uf("2020-05-05", "28000.00")

        val result = UfReajuste.convert(BigDecimal("1234"), day, day)

        assertThat(result.adjustedAmount.toDouble()).isWithin(0.5).of(1234.0)
        assertThat(result.factor.toDouble()).isWithin(1e-9).of(1.0)
        assertThat(result.variationPct.toDouble()).isWithin(1e-9).of(0.0)
        assertThat(result.annualisedPct.toDouble()).isWithin(1e-9).of(0.0)
        assertThat(result.days).isEqualTo(0L)
    }

    @Test
    fun `converting forward and back returns the original amount`() {
        val a = uf("1995-06-15", "12000.00")
        val b = uf("2020-06-15", "28000.00")

        val forward = UfReajuste.convert(BigDecimal("100000"), a, b)
        val back = UfReajuste.convert(forward.adjustedAmount, b, a)

        assertThat(back.adjustedAmount.toDouble()).isWithin(1.0).of(100000.0)
    }

    @Test
    fun `going backwards in time shrinks the amount and reports negative variation`() {
        val result = UfReajuste.convert(
            amount = BigDecimal("30000"),
            from = uf("2026-09-05", "40880.36"),
            to = uf("1990-01-01", "5435.28"),
        )

        assertThat(result.adjustedAmount.toDouble()).isLessThan(30000.0)
        assertThat(result.variationPct.signum()).isEqualTo(-1)
        assertThat(result.days).isEqualTo(-13396L)
    }

    @Test
    fun `the annual rate compounds back to the factor`() {
        val result = UfReajuste.convert(
            amount = BigDecimal("1000"),
            from = uf("2000-01-01", "15000.00"),
            to = uf("2020-01-01", "28000.00"),
        )
        val years = result.days / 365.25
        val rebuilt = Math.pow(1 + result.annualisedPct.toDouble() / 100, years)

        assertThat(rebuilt).isWithin(0.01).of(result.factor.toDouble())
    }

    /** Day precision is the point: consecutive days do not give the same answer. */
    @Test
    fun `two dates one day apart give different results`() {
        val base = uf("2026-09-01", "40875.09")
        val a = UfReajuste.convert(BigDecimal("1000000"), base, uf("2026-09-04", "40879.04"))
        val b = UfReajuste.convert(BigDecimal("1000000"), base, uf("2026-09-05", "40880.36"))

        assertThat(b.adjustedAmount).isGreaterThan(a.adjustedAmount)
    }

    @Test
    fun `large amounts do not drift from double rounding`() {
        val result = UfReajuste.convert(
            amount = BigDecimal("1000000000"),
            from = uf("2020-01-01", "28000.00"),
            to = uf("2020-01-02", "28001.00"),
        )

        // Computed from the factor once, not by rounding UF units and back.
        assertThat(result.adjustedAmount.toDouble())
            .isWithin(1.0).of(1_000_000_000.0 * 28001 / 28000)
    }

    @Test
    fun `a non-positive UF value is refused rather than dividing by zero`() {
        assertThrows(IllegalArgumentException::class.java) {
            UfReajuste.convert(BigDecimal("1000"), uf("2020-01-01", "0"), uf("2021-01-01", "1"))
        }
        assertThrows(IllegalArgumentException::class.java) {
            UfReajuste.convert(BigDecimal("1000"), uf("2020-01-01", "1"), uf("2021-01-01", "-1"))
        }
    }
}
